src/app/core/nav-menu/nav-menu.service.ts

```diff
@@ -0,0 +1,58 @@
+import { Injectable } from '@angular/core';
+import { select, Store } from '@ngrx/store';
+import { filter, map, withLatestFrom } from 'rxjs/operators';
+import { Observable, of } from 'rxjs';
+
+import { Route, Subroute } from './models';
+import { appNavRoutes } from './routes';
+import { routerSelectors, State } from '../../root-store';
+
+
+@Injectable()
+export class NavMenuService {
+  constructor(private store: Store<State>) {
+  }
+
+  public getRoutes(): Observable<Route[]> {
+    return of(appNavRoutes);
+  }
+
+  public getSubroutes(): Observable<Subroute[]> {
+    return this.getCurrentRoute().pipe(
+      map(route => route.subroutes)
+    );
+  }
+
+  public getCurrentRoute(): Observable<Route> {
+    return this.getCurrentSubroute().pipe(
+      filter(subroute => !!subroute),
+      withLatestFrom(this.getRoutes()),
+      map(([subroute, routes]) => routes.find(route => route.id === subroute.routeId))
+    );
+  }
+
+  private getCurrentSubroute(): Observable<Subroute> {
+    return this.getCurrentSubroutePath().pipe(
+      map(path => {
+        const subroutes = this.getAllSubroutes();
+        return subroutes.find(subroute => subroute.path === path)
+      })
+    );
+  }
+
+  private getCurrentSubroutePath(): Observable<string> {
+    return this.store.pipe(
+      select(routerSelectors.getUrl),
+      filter(url => url != null),
+      map(url => {
+        const matches = url.match(/^\/[A-Za-z-]*/);
+        return matches[0];
+      }),
+    );
+  }
+
+  private getAllSubroutes(): Subroute[] {
+    const subroutesArrays = appNavRoutes.map(route => route.subroutes);
+    return [].concat.apply([], subroutesArrays);
```

Comment:
This construction a bit unreadable. Maybe it is better to use `reduce` instead?

```
appNavRoutes.map(route => route.subroutes)
  .reduce((x, acc) => [...acc, ...x], [])
```

Reply:
Okay. But what do you think about extract `[].concat.apply([], subroutesArrays)` to a pure function with `flattenArray` name. Or better use a reduce approach. I just wondering for myself.